File: output/playwright/2026-08-13-62f8de8-alex-final/audit.py
from __future__ import annotations

import asyncio
import json
from pathlib import Path
import time

from playwright.async_api import async_playwright


ROOT = Path(__file__).resolve().parents[1] / "2026-08-13-ce0f9e6-alex-final"
ROOT.mkdir(parents=True, exist_ok=True)
BASE = "https://spreadarbitrage.ink"
TOKEN_POINTER = Path("/tmp/spreadboard-qa-session-path")


async def main() -> None:
    token_path = Path(TOKEN_POINTER.read_text().strip())
    token = token_path.read_text().strip()
    results: list[dict] = []
    failures: list[str] = []
    cases = [
        ("markets", "/markets"),
        ("funding", "/funding"),
        ("rankings", "/rankings"),
        ("charts", "/charts?token=GUA"),
        ("account", "/account"),
    ]
    async with async_playwright() as pw:
        browser = await pw.chromium.launch(
            headless=True,
            executable_path="/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
        )
        for viewport, size in (
            ("desktop", {"width": 1440, "height": 1000}),
            ("mobile", {"width": 390, "height": 844}),
        ):
            context = await browser.new_context(viewport=size, device_scale_factor=1)
            await context.add_cookies(
                [{
                    "name": "spreadboard_session",
                    "value": token,
                    "domain": "spreadarbitrage.ink",
                    "path": "/",
                    "secure": True,
                    "httpOnly": True,
                    "sameSite": "Lax",
                }]
            )
            for name, path in cases:
                page = await context.new_page()
                errors: list[str] = []
                page.on("pageerror", lambda exc, errors=errors: errors.append(str(exc)))
                page.on(
                    "console",
                    lambda msg, errors=errors: errors.append(f"console:{msg.text}")
                    if msg.type == "error"
                    else None,
                )
                started = time.perf_counter()
                response = await page.goto(BASE + path, wait_until="domcontentloaded", timeout=60_000)
                dom_ms = round((time.perf_counter() - started) * 1000, 1)
                await page.wait_for_timeout(2500)
                body = await page.locator("body").inner_text()
                h1 = await page.locator("h1").first.inner_text() if await page.locator("h1").count() else ""
                overflow = await page.evaluate(
                    "Math.max(0, document.documentElement.scrollWidth - document.documentElement.clientWidth)"
                )
                lowered = body.lower()
                invalid = [needle for needle in ("traceback", "internal server error", "nan%", "infinity%") if needle in lowered]
                state: dict[str, object] = {}
                if name == "rankings":
                    headers = " ".join(await page.locator("th").all_inner_texts()).lower()
                    state = {
                        "rows": await page.locator("tbody tr").count(),
                        "has1d": "settled 24h" in headers,
                        "has7d": "settled 7d" in headers,
                        "has30d": "settled 30d" in headers,
                        "guard_markers_in_top": await page.locator("tbody tr").evaluate_all(
                            "rows => rows.slice(0, 10).filter(row => row.innerText.includes('?')).length"
                        ),
                    }
                elif name == "charts":
                    for _ in range(80):
                        if await page.locator("[data-chart-long] option").count() > 1:
                            break
                        await page.wait_for_timeout(250)
                    options = " ".join(await page.locator("option").all_inner_texts()).lower()
                    long_values = await page.locator("[data-chart-long] option").evaluate_all(
                        "options => options.slice(1).map(option => option.value)"
                    )
                    short_values = await page.locator("[data-chart-short] option").evaluate_all(
                        "options => options.slice(1).map(option => option.value)"
                    )
                    if long_values and short_values:
                        long_value = next((value for value in long_values if "okx dex" in value.lower()), long_values[0])
                        short_value = next((value for value in short_values if value != long_value and "futures" in value.lower()), "")
                        if not short_value:
                            short_value = next((value for value in short_values if value != long_value), "")
                        await page.locator("[data-chart-long]").select_option(long_value)
                        await page.locator("[data-chart-short]").select_option(short_value)
                    state = {
                        "gua": "gua" in lowered,
                        "dex": "dex" in options,
                        "futures": "futures" in options,
                        "ready": not await page.locator("[data-chart-create]").is_disabled(),
                    }
                elif name == "account":
                    state = {
                        "edit": "edit" in lowered,
                        "delete": "delete" in lowered,
                        "movement_pnl": "movement pnl" in lowered,
                        "settled_funding": "settled funding" in lowered,
                        "total_pnl": "total pnl" in lowered,
                    }
                screenshot = ROOT / f"{viewport}-{name}.png"
                await page.screenshot(path=str(screenshot), full_page=True)
                item = {
                    "viewport": viewport,
                    "name": name,
                    "status": response.status if response else None,
                    "final": page.url,
                    "dom_content_loaded_ms": dom_ms,
                    "errors": errors,
                    "state": state,
                    "screenshot": str(screenshot),
                    "title": await page.title(),
                    "h1": h1,
                    "overflow": overflow,
                    "invalid": invalid,
                    "chars": len(body),
                }
                results.append(item)
                if item["status"] != 200 or errors or overflow or invalid or "/login" in page.url:
                    failures.append(f"{viewport}:{name}:transport_or_render")
                if name == "rankings" and (
                    state["rows"] < 50 or not state["has1d"] or not state["has7d"] or not state["has30d"]
                ):
                    failures.append(f"{viewport}:{name}:ranking_columns")
                if name == "charts" and not all(state.values()):
                    failures.append(f"{viewport}:{name}:gua_route_picker")
                if name == "account" and not all(state.values()):
                    failures.append(f"{viewport}:{name}:portfolio_controls")
                await page.close()
            await context.close()
        await browser.close()
    report = {"generated_at": "2026-08-13", "revision": "ce0f9e6", "results": results, "failures": failures}
    (ROOT / "report.json").write_text(json.dumps(report, indent=2) + "\n")
    print(json.dumps(report, indent=2))
    raise SystemExit(1 if failures else 0)


if __name__ == "__main__":
    asyncio.run(main())

File: output/playwright/2026-08-27-unified-final/audit.js
"use strict";

const fs = require("node:fs");
const path = require("node:path");
const { chromium } = require("playwright");

const base = "https://spreadarbitrage.ink";
const session = process.env.SPREADBOARD_AUDIT_SESSION;
const exactPair = process.env.SPREADBOARD_AUDIT_PAIR || "";
const output = __dirname;

async function inspect(page, viewport) {
  const errors = [];
  page.on("pageerror", error => errors.push(`page:${error.message}`));
  page.on("console", message => {
    if (message.type() === "error") errors.push(`console:${message.text()}`);
  });
  await page.context().addCookies([{
    name: "spreadboard_session",
    value: session,
    url: base,
    httpOnly: true,
    secure: true,
    sameSite: "Lax",
  }]);
  const started = Date.now();
  const response = await page.goto(`${base}/funding?farm=futures-futures&rank=now&q=SIREN`, {
    waitUntil: "domcontentloaded",
    timeout: 90_000,
  });
  await page.waitForTimeout(700);
  const funding = await page.evaluate(() => {
    const text = document.body.innerText;
    const links = [...document.querySelectorAll('a[href^="/pair/"]')];
    const sirenPair = links.find(link => {
      const row = link.closest("article, tr, section, div");
      return (row?.innerText || link.innerText).includes("SIREN");
    }) || links[0];
    return {
      bodyChars: text.length,
      researchOnly: text.includes("Research only"),
      windows: ["Now", "24h", "7d", "30d"].filter(label => text.includes(label)),
      exactValues: ["+0.16%", "+0.85%", "+0.10%"].filter(value => text.includes(value)),
      invalid: ["undefined", "NaN", "[object Object]"].filter(value => text.includes(value)),
      overflow: Math.max(0, document.documentElement.scrollWidth - innerWidth),
      pairHref: sirenPair?.getAttribute("href") || null,
    };
  });
  await page.screenshot({ path: path.join(output, `${viewport}-funding-siren.png`), fullPage: true });
  let pair = null;
  if (exactPair || funding.pairHref) {
    const pairResponse = await page.goto(`${base}${exactPair || funding.pairHref}`, {
      waitUntil: "domcontentloaded",
      timeout: 90_000,
    });
    await page.waitForTimeout(500);
    pair = await page.evaluate(() => {
      const text = document.body.innerText;
      return {
        statusText: document.querySelector("h1")?.innerText || "",
        routeReturns: text.includes("Route returns"),
        windows: ["Now", "24h", "7d", "30d"].filter(label => text.includes(label)),
        exactValues: ["+0.16%", "+0.85%", "+0.10%"].filter(value => text.includes(value)),
        invalid: ["undefined", "NaN", "[object Object]"].filter(value => text.includes(value)),
        overflow: Math.max(0, document.documentElement.scrollWidth - innerWidth),
      };
    });
    pair.httpStatus = pairResponse.status();
    await page.screenshot({ path: path.join(output, `${viewport}-pair-detail.png`), fullPage: true });
  }
  return {
    viewport,
    status: response.status(),
    elapsedMs: Date.now() - started,
    funding,
    pair,
    errors,
  };
}

(async () => {
  fs.mkdirSync(output, { recursive: true });
  const browser = await chromium.launch({
    headless: true,
    executablePath: process.env.PLAYWRIGHT_EXECUTABLE_PATH || undefined,
  });
  const results = [];
  try {
    for (const [name, viewport] of [["desktop", { width: 1440, height: 1000 }], ["mobile", { width: 390, height: 844 }]]) {
      const context = await browser.newContext({ viewport, colorScheme: "dark" });
      results.push(await inspect(await context.newPage(), name));
      await context.close();
    }
  } finally {
    await browser.close();
  }
  const failures = results.filter(result =>
    result.status !== 200 || result.errors.length || result.funding.researchOnly ||
    result.funding.windows.length !== 4 || result.funding.invalid.length || result.funding.overflow > 2 ||
    !result.pair || result.pair.httpStatus !== 200 || !result.pair.routeReturns ||
    result.pair.windows.length !== 4 || result.pair.exactValues.length !== 3 ||
    result.pair.invalid.length || result.pair.overflow > 2
  );
  console.log(JSON.stringify({ results, failures }, null, 2));
  if (failures.length) process.exitCode = 1;
})();
